aptos-move/aptos-vm/src/parallel_executor/mod.rs

```diff
@@ -58,6 +64,47 @@ impl PTransactionOutput for AptosTransactionOutput {
     }
 }
 
+static RAYON_EXEC_POOL: Lazy<rayon::ThreadPool> = Lazy::new(|| {
+    rayon::ThreadPoolBuilder::new()
+        .num_threads(num_cpus::get())
+        .thread_name(|index| format!("parallel_executor_{}", index))
+        .build()
+        .unwrap()
+});
+
+struct LruCache16ShardWrapper {
+    state: LruCache<[u8; 32], u64>,
+}
+
+impl LruCache16ShardWrapper {
+    fn new(cache_size_per_shard: usize) -> LruCache16ShardWrapper {
+        LruCache16ShardWrapper {
+            state: LruCache::new(cache_size_per_shard as u64),
+        }
+    }
+
+    pub fn get_key(&self, item: &Transaction) -> [u8; 32] {
+        let bytes = to_bytes(item).unwrap();
+        let mut hasher = DefaultHasher::new(b"CacheTesting");
+        hasher.update(&bytes);
+        hasher.finish().get_bytes()
+    }
+
+    /// return whether the item exists (cache hit).
+    /// NOTE: false cache miss is possible.
+    pub fn insert(&self, item: &Transaction) -> bool {
+        let nonce = thread_rng().gen::<u64>();
+        let entry = self.state.get_or_init(self.get_key(item), 1, |_e| nonce);
```

Comment:
oh this nonce business :-)
definitely doesn't look like a good library, given the APIs are so limited, shards fixed.